### ISSUE TITLE
be more resilient against small changes in the Requirements string

### ISSUE THREAD

**Issue by rseuster** [hep-gc/cloudscheduler#275]
With the latest changes, we see lots of foreign jobs now.
The problem is now that Frank slightly changed the Requirements string.
```
[seuster@csv2a ~]$ condor_q -af:r Requirements | uniq -c
     71 (group_name =?= "atlas") && (TARGET.Arch == "X86_64") && (TARGET.OpSys == "LINUX") && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && ((TARGET.HasFileTransfer) || (TARGET.FileSystemDomain == MY.FileSystemDomain))
    470 (group_name =?= "atlas") && (target_alias =?= "CA-IAAS-T3_UCORE") && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && (TARGET.HasFileTransfer)
    649 (group_name =?= toLower("ATLAS")) && (target_alias =?= "CA-IAAS-T3_UCORE") && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && (TARGET.HasFileTransfer)
     14 (group_name == "atlas") && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && (TARGET.HasFileTransfer)
```

Note that now, we have (group_name == "atlas") whereas it was (group_name =?= "atlas") before.

On python level this now comes back differently, therefore jobs are not recognized properly as belonging to this group; jpoll is basically doing what the csjob poller is doing when getting job class ads:
```
[seuster@csv2a ~]$ python3 ./jpoll.py | sort | uniq -c
     14 (group_name == "atlas") && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && (TARGET.HasFileTransfer)
    470 (group_name is "atlas") && (target_alias is "CA-IAAS-T3_UCORE") && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && (TARGET.HasFileTransfer)
     71 (group_name is "atlas") && (TARGET.Arch == "X86_64") && (TARGET.OpSys == "LINUX") && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && ((TARGET.HasFileTransfer) || (TARGET.FileSystemDomain == MY.FileSystemDomain))
    649 (group_name is toLower("ATLAS")) && (target_alias is "CA-IAAS-T3_UCORE") && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && (TARGET.HasFileTransfer)
```
Here
https://github.com/hep-gc/cloudscheduler/blob/dev/data_collectors/condor/csjobs.py#L214

we are matching against: "group_name is" but that now is "group_name ==".

I am wondering if there's a possibility to have condor do the test rather than is interpreting the Requirements ad.

**Comment by rseuster** [comment]
I modified the Requirements string to read again (group_name =?= "atlas") because VMs were retiring, apparently no more idle jobs in the queue....

**Comment by colsond** [comment]
I've made csjobs slightly more resilient with commit 03773d1a091f93fcacf7b242ee7255bdab555898 as stated in the commit it should be able to properly parse "is", "==", and "=?=" as well as the toLower condor function. csjobs always takes the result and stores it as lowercase.

**Comment by rseuster** [comment]
I think I got it. I have modified one job to have this Requirements classad:
```
[seuster@csv2a ~]$ condor_q -af:r requirements ClusterId | grep toL
(group_name == toLower("atlas")) && (target_alias =?= toLower("CA-IAAS-T3_UCORE")) && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && (TARGET.HasFileTransfer) 694507
```
and jpoll.py for this job gives:
```
(group_name == toLower("atlas")) && (target_alias is toLower("CA-IAAS-T3_UCORE")) && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && (TARGET.HasFileTransfer)
Got this : requirements = (group_name == toLower("atlas")) && (target_alias is toLower("CA-IAAS-T3_UCORE")) && (TARGET.Cpus >= RequestCpus) && (TARGET.Memory >= RequestMemory) && (TARGET.HasFileTransfer) evaluates as Undefined 
    [
        RequestDisk = 15; 
        ClusterId = 694507; 
        JobStatus = 1; 
        Requirements = (group_name == "atlas") && (target_alias is "ca-iaas-t3_ucore") && (TARGET.Cpus >= 1) && (TARGET.Memory >= 3000) && (TARGET.HasFileTransfer); 
        ServerTime = 1576719072; 
        GlobalJobId = "csv2a.heprc.uvic.ca#694507.0#1576706212"; 
        Owner = "autopyfactory"; 
        RequestCpus = 1; 
        DiskUsage = 15; 
        User = "autopyfactory@csv2a.heprc.uvic.ca"; 
        QDate = 1576706185; 
        JobPrio = 0; 
        RequestMemory = 3000; 
        ProcId = 0; 
        EnteredCurrentStatus = 1576706235
    ]..
```
Here's the code:
```
#!/usr/bin/python3

import htcondor
import classad

condor_host='csv2a.heprc.uvic.ca'
coll = htcondor.Collector(condor_host)
scheddAd = coll.locate(htcondor.DaemonTypes.Schedd, condor_host)
condor_session = htcondor.Schedd(scheddAd)

job_attributes = ["TargetClouds", "JobStatus", "RequestMemory", "GlobalJobId", "HoldReason",
                  "RequestDisk", "RequestCpus", "RequestScratch", "RequestSwap", "Requirements",
                  "JobPrio", "ClusterId", "ProcId", "User", "VMInstanceType", "VMNetwork",
                  "VMImage", "VMKeepAlive", "VMMaximumPrice", "VMUserData", "VMJobPerCore", "Owner",
                  "EnteredCurrentStatus", "QDate", "HoldReasonCode", "HoldReasonSubCode", 
                  "LastRemoteHost", "TargetAlias"]

job_list = condor_session.xquery(
    projection=job_attributes
)

for job_ad in job_list:
    job_dict = dict(job_ad)
    if "Requirements" in job_dict:
        print (job_dict['Requirements'])
        m="requirements = " + str(job_dict['Requirements']) 
        et1=classad.ExprTree(m).eval("requirements")
        #jd={"requirements": '(group_name is "atlas")'}
        ca1=classad.ClassAd(job_dict)
        et2=ca1.flatten(job_dict)
        print ("Got this : %s evaluates as %s %s.." %(m,str(et1),et2))
```

I didn't check how expensive this code is exactly, but for the jobs currently in csv2 it's pretty fast.

**Comment by rseuster** [comment]
Under "Predefined Functions" one can see all the available functions, currently over 50:
https://htcondor.readthedocs.io/en/latest/misc-concepts/classad-mechanism.html

**Comment by colsond** [comment]
Fixed in commit 5716b3cfa81a760c47157e39630b719dbaaeede8
